### PR TITLE
Backup add

### PR DESCRIPTION
**Pull request by aswietlik**
Not implemented yet (to be done in next PRs):
- "OneTime" backups
- default values - name and storage
- create separate pvc for each backup
- list of created backups as cluster events

### CODE REVIEW

File: pkg/apis/cr/v1/backup.go [hunk 1]
```diff
@@ -18,6 +19,9 @@ type MySQLBackup struct {
 
 // MySQLBackupSpec stores the properties of a MySQL cluster backup.
 type MySQLBackupSpec struct {
+	Cluster string            `json:"cluster"`
+	Time    string            `json:"time"`
+	Storage resource.Quantity `json:"storage"`
```
Comment (earlgreyz) [inline pkg/apis/cr/v1/backup.go]: We should keep a list of backups already created with this Spec to easily restore a cluster from specific file.

For example keep a list of such entries (update during cronjob)
```go
type MySQLBackupInstance {
        CreatedAt string `json:"createdAt"`
        Filename  string `json:"filename"`
        Status    string `json:"status",omitempty`
}

File: operator/backup/operator.go [hunk 2]
```diff
@@ -12,17 +24,91 @@ type Operator interface {
...
+	}
+
 	return nil
 }
+
+func (b *backupOperator) createPVC(backup *crv1.MySQLBackup) error {
+	pvcInterface := b.clientset.CoreV1().PersistentVolumeClaims(backup.Namespace)
+	pvc, err := pvcForBackup(backup)
+	if err != nil {
+		logging.LogBackup(backup).Debug("PVC error: " + fmt.Sprint(err))
```
Comment (earlgreyz) [inline operator/backup/operator.go]: Just return an error and move the logging to the controller.

File: operator/backup/operator.go [hunk 3]
```diff
@@ -12,17 +24,91 @@ type Operator interface {
...
+	}
+
+	return nil
+}
+
+func (b *backupOperator) createCronJob(backup *crv1.MySQLBackup) error {
+	cronJobInterface := b.clientset.BatchV1beta1().CronJobs(backup.Namespace)
+	cronJob, err := cronJobForBackup(backup)
+	if err != nil {
+		logging.LogBackup(backup).Debug("Cron job error: " + fmt.Sprint(err))
```
Comment (earlgreyz) [inline operator/backup/operator.go]: Just return an error and move the logging to the controller.

File: operator/backup/operator.go [hunk 4]
```diff
@@ -12,17 +24,91 @@ type Operator interface {
...
+
+	logging.LogBackup(backup).Debug("Creating PVC.")
+	err = b.createPVC(backup)
+	if err != nil {
+		return err
+	}
+
+	logging.LogBackup(backup).Debug("Creating cron job.")
+	err = b.createCronJob(backup)
+	if err != nil {
```
Comment (earlgreyz) [inline operator/backup/operator.go]:
```go
logging.LogBackup(backup).WithField("error", err).Warn("Reverting PVC creation.")
```

Comment (aswietlik) [reply operator/backup/operator.go]: Done.

Comment (aswietlik) [reply operator/backup/operator.go]: Done.

File: operator/backup/operator.go [hunk 5]
```diff
@@ -12,17 +23,90 @@ type Operator interface {
...
+		// Cleanup - remove already created PVC
+		logging.LogBackup(backup).WithField("error", err).Warn("Reverting PVC creation.")
+		b.removePVC(backup)
```
Comment (earlgreyz) [inline operator/backup/operator.go]: Like in `Cluster operator` if an error during removal occurs we want to let the user know:

```go
removeErr := removePVC(cluster)
return errors.NewAggregate([]error{err, removeErr})
```

File: pkg/apis/cr/v1/backup.go (hunk 1 above)
Comment (aswietlik) [reply pkg/apis/cr/v1/backup.go]: We'll provide it as a list of events. To be done in next pull requests.

**Pull request merged**